src/avram/bulk_upsert.cr

```diff
@@ -0,0 +1,70 @@
+class Avram::BulkUpsert(T)
+  def initialize(@records : Array(T), @column_names : Array(Symbol))
+    @records = set_timestamps(records)
+  end
+
+  def statement
+    [
+      "INSERT INTO #{table}(#{fields})",
+      "VALUES #{value_placeholders}",
+      "ON CONFLICT (#{conflicts}) DO UPDATE SET #{updates}",
+      "RETURNING #{returning}",
+    ].join(" ")
+  end
+
+  private def conflicts
+    @column_names.join(", ")
+  end
+
+  private def set_timestamps(collection)
+    collection.map do |record|
+      record.created_at.value ||= Time.utc if record.responds_to?(:created_at)
+      record.updated_at.value = Time.utc if record.responds_to?(:updated_at)
+      record
+    end
+  end
+
+  private def table
+    @records.first.table_name
+  end
+
+  private def updates
+    update_keys = @records.first.insert_values.keys
+
+    (update_keys - [:created_at]).map do |column|
+      "#{column}=EXCLUDED.#{column}"
+    end.join(", ")
+  end
+
+  private def returning
+    T.column_names.join(", ")
+  end
+
+  private def fields
+    @records.first.insert_values.keys.map do |key|
+      <<-TEXT
+      "#{key}"
+      TEXT
+    end.join(", ")
+  end
+
+  def args
+    @records.flat_map do |record|
+      record.insert_values.values
+    end
+  end
+
+  private def value_placeholders(record)
+    record.insert_values.map_with_index(1) do |_value, index|
+      "$#{index}"
+    end.join(", ")
+  end
+
+  private def value_placeholders
+    i = 0
+
+    @records.map do |record|
+      "($#{i += 1}, $#{i += 1}, $#{i += 1}, $#{i += 1})"
```

Comment:
🤷

Reply:
We were testing with 4 columns, this will need to be dynamic based off of columns.count

Reply:
This was to test a theory about the right way to pass bind args -- will be removed, since bind args seem like a non-starter due to the 32k var cap.

Reply:
if the cap is 32k variables per query, could you do some basic math and split a statement into multiple queries? I would really like to avoid jumping into sql injection land.

Reply:
Something like:

```crystal
field_count = record.columns.count
records.each_slice(32767 / field_count) do |record_set|
   upsert(record_set)
end
```

I don't think I'd have a problem with raising an exception if you hit this maximum and just requiring the API to take a max of rows where 32767 > field_count * row_count.

Reply:
robacarp example, for prosperity: https://github.com/luckyframework/avram/pull/789#discussion_r783094748

Reply:
Went with an approach that won't limit our record count.